### ISSUE TITLE
adding signal-android devices

### ISSUE THREAD

**Issue by Trolldemorted** [AsamK/signal-cli#20]
is it possible to add signal for android as a non-master device?

i know that the android devices cannot display a tsdevice link, but since it is nothing but an encoded public key (?) i guess i can construct it manually with a self-compiled signal android.

did you try it yet and do you know any imminent problems?

**Comment by AsamK** [comment]
It's possible, if you implement it yourself.  You need to build Signal-Android with a libsignal-service-java version, that contains the following PR: WhisperSystems/libsignal-service-java#21 .
And then you have modify Signal-Android to create the tsdevice link, call the appropriate provisioning methods in libsignal-service-java and make Signal-Android send its deviceId in all server requests. here's the commit that adds provisioning as non-master to signal-cli: 33956bde62d4fc7de5325bf3bb7be3b323863442
Also contact and group synchronization need to be implemented.

So I guess it's a lot of work and I'm not sure if it's going to be merged upstream.
I haven't attempted it and anyway, it's not a signal-cli issue.

**Comment by Trolldemorted** [comment]
Would you be willing to help, though?

I have managed to make Signal-Android produce tsdevice links, finish the device link, and append its received deviceid to all requests so far.

sending and receiving personal and group messages is working fine, but the initial group and contact synchronization is troubling me. When i send a sync-request like you did [here](https://github.com/AsamK/signal-cli/blob/master/src/main/java/org/asamk/signal/Manager.java#L625), the api throws a MismatchedDevicesException:

```
07-13 00:55:10.536 3398-3427/org.thoughtcrime.securesms W/SignalServiceMessageSender: org.whispersystems.signalservice.internal.push.exceptions.MismatchedDevicesException
  at org.whispersystems.signalservice.internal.push.PushServiceSocket.makeRequest(PushServiceSocket.java:538)
  at org.whispersystems.signalservice.internal.push.PushServiceSocket.sendMessage(PushServiceSocket.java:207)
  at org.whispersystems.signalservice.api.SignalServiceMessageSender.sendMessage(SignalServiceMessageSender.java:358)
  at org.whispersystems.signalservice.api.SignalServiceMessageSender.sendMessage(SignalServiceMessageSender.java:213)
  at org.thoughtcrime.securesms.jobs.GroupSyncRequestJob.onRun(GroupSyncRequestJob.java:48)
  at org.thoughtcrime.securesms.jobs.MasterSecretJob.onRun(MasterSecretJob.java:18)
  at org.whispersystems.jobqueue.JobConsumer.runJob(JobConsumer.java:76)
  at org.whispersystems.jobqueue.JobConsumer.run(JobConsumer.java:46)
```

could you give me a hint what a MismatchedDevicesException is supposed to indicate? I cannot catch it, since jobs run in different threads, i suppose.

**Comment by AsamK** [comment]
Nice work!
I think a mismatched device exception occurs, when the client tries to send a message, that is not encrypted for all of the linked devices or for a device that is not linked anymore. You could check if the send methods called for the sync messages are the same as in signal-cli and also if the deviceId is always set for the sync messages.
If you push your code to your fork, I could have a look

**Comment by Trolldemorted** [comment]
I cleaned up the source code as much as i could, here is the commit: https://github.com/Trolldemorted/Signal-Android/commit/8ac7f5208fbfad3a75e949471f7733a9f41ef7d4

Don't forget to build against https://github.com/WhisperSystems/libsignal-service-java/pull/21 - i did not want to waste time on figuring how gradle/maven work, so i dropped my compiled jars manually into my gradle cache. If you know an easy and better solution, please elaborate :)

**Comment by AsamK** [comment]
The exception is raised because Signal-Android creates a session for its own deviceId, and the server doesn't like that.
I've added a workaround in libsignal-service-java, which should prevent the MismatchedDevicesException: https://github.com/AsamK/libsignal-service-java/commit/f5c4e7c0aedd60d3d13814e3c9293e6054f6f4e5

You can use `./gradlew installArchives` to install libsignal-service-java to your local maven repository which is picked up by Signal-Android with the mavenLocal() entry.

**Comment by Trolldemorted** [comment]
Thanks a lot for your quick help!

Since i use your libsignal-service no more MismatchedDevicesExceptions are popping up - you might want to overload setAccountAttributes and verifyAccountWithCode though, since the RegistrationService, RegistrationProgressActivity and RefreshAttributesJob classes won't compile due to calls with missing parameters. Do you think you will manage to convice OWS to merge your fixes into their master branch?

As for signal-cli:
when receiving a contact sync request, signal-cli prints out an appropriate message:
```
Envelope from: +491XXXXXXXXX (device: 4)
Timestamp: 1468456960583
Received a sync message
Received sync request
 - contacts request
```
Naturally it would be great if it would respond to it (either automatically or by explicit command - i guess calling pushContacts after loading the config still works, so at least i got a workaround for now).

when receiving a group sync request, signal-cli prints a NullPointerException:
```
Exception in thread "main" java.lang.NullPointerException
    at org.whispersystems.signalservice.internal.push.SignalServiceProtos$GroupDetails$Avatar$Builder.setContentType(SignalServiceProtos.java:11333)
    at org.whispersystems.signalservice.api.messages.multidevice.DeviceGroupsOutputStream.writeGroupDetails(DeviceGroupsOutputStream.java:41)
    at org.whispersystems.signalservice.api.messages.multidevice.DeviceGroupsOutputStream.write(DeviceGroupsOutputStream.java:17)
    at org.asamk.signal.Manager.sendGroups(Manager.java:1017)
    at org.asamk.signal.Manager.receiveMessages(Manager.java:809)
    at org.asamk.signal.Main.main(Main.java:360)
```
but i cannot yet tell whether my request is malformed or signal-cli is crashing on a correct request.

**Comment by AsamK** [comment]
Probably not, so far none of my libsignal-service-java PRs have been merged or even been comment on.

signal-cli should already send the sync contacts after receiving a contact sync request.
The same happens for group requests. The error you get there was caused when sending the sync groups, by an avatar image, of which java couldn't detect the content type. This should be fixed in the master branch now.

**Comment by Trolldemorted** [comment]
I just tested it (or rather, i had to, new OS) with my old signal-cli master instance and a new signal desktop instance.

After adding the new device, S-D says its trying to sync contacts&groups, i execute S-C -receive, output looks good but no contacts or groups pop up in S-D:
```
Envelope from: +491XXXXXXXXX (device: 5)
Timestamp: 1468537314505
Received a sync message
Received sync request
 - contacts request


Envelope from: +491XXXXXXXXX (device: 5)
Timestamp: 1468537316173
Received a sync message
Received sync request
 - groups request
```

Then i noticed that my contact store is empty, but i *definitely* did not do that manually. I checked my old backup, and even there the contact store is gone. Unfortunately i do not know when and on which version this occured.

Nevertheless, the group store is intact, but S-D does not display any groups. S-D printed an error to the console, though:

```https://textsecure-service-ca.whispersystems.org:4433/v1/messages/+491628396072 Failed to load resource: the server responded with a status of 409 (Conflict)````



When i use S-D's function to import contacts&groups from the master device, S-C -receive prints the requests to the console, S-D prints ```group sync``` and dispatches some HTTP-GET requests for attachments (i guess for the group icons, the amount seems to fit) but does not show any groups.
